compiler/rustc_lexer/src/cursor.rs

```diff
@@ -91,7 +91,17 @@ impl<'a> Cursor<'a> {
         // It was tried making optimized version of this for eg. line comments, but
         // LLVM can inline all of this and compile it down to fast iteration over bytes.
         while predicate(self.first()) && !self.is_eof() {
-            self.bump();
+            #[cfg(debug_assertions)]
+            {
+                self.bump();
+            }
+
+            #[cfg(not(debug_assertions))]
+            {
+                // SAFETY: we checked it's not EOF therefore there must be
+                // at least 1 char present.
+                unsafe { self.chars.advance_by(1).unwrap_unchecked() };
```

Comment:
YMMV: Given that we've seen `unwrap_unchecked` not actually remove the branches on slice iterators, consider implementing <https://stdrs.dev/nightly/x86_64-unknown-linux-gnu/core/iter/traits/unchecked_iterator/trait.UncheckedIterator.html> and having this use that.  (And maybe have `next` call it too.)

Reply:
It does appear to help here.

https://rust.godbolt.org/z/3eod9e6je